Enable JoinMulticastGroupAsync on UWP for ports that are already in use

1iveowl opened the pull request:
Regarding issue #85. I tried to make a PR towards the dev branch and not the 1.2.2 branch as I could see that the dev branch is on VS 2015.  

See issue #85 for details. It's not much really. I basically added a UWP Project and added 3 lines of code to _UdpSocketMulticastClient.cs_ in _Sockets.Implementation.WinRT_.

Unfortunately I had to make some more changes to the dev branch for it to compile. Specifically it looks like `WrapNativeSocketExceptionsAsTask` was reverted back in January for the dev branch. Not sure what this is about.  

I'm not sure where the two .txt files and the csproj.bak came from, but I could have been because I updated the Nugets.

1iveowl committed:
Enable JoinMulticastGroupAsync on UWP for ports that are already in use

rdavisau commented:
Hey - thanks for this! Apologies for the delay - I was away over the weekend.

A few small things:

- Unfortunately, right now this should be targeted against the `v2.0.0-pre` branch - apologies as that is not specified anywhere. The `WrapNativeSocketExceptions` issue should not be present in the 2.0.0 branch. With github right now I believe you need to open a new PR to change the target.
- Don't worry about committing the weird files - the `.bak` and `.txt` files can be excluded

As for the change itself:

- As this is a cross-platform library, we need this behavior to be consistent across all platforms, or note with a caveat that it's possible on certain platforms where that applies. I think it should be possible on all platforms here, but may require for compiler directives for the WinRT/UWP side - I can see you added a dedicated UWP project - does that mean the `MulticastOnly` property (what an odd name!) is not available in WinRT? If you investigate WinRT side, I will find the solution for the the .NET side. 

- I don't think sharing ports should be the default option. Instead, we should add an optional parameter to `JoinMulticastGroupAsync` with `false` as the default value, to control whether we can reuse a used multicast port. That requires the signature to change in `IUdpMulticastReceiver` as well. 

Let me know if you are happy to make these changes and retarget against the 2.0.0 branch and I will work on the .NET side.

1iveowl commented:
It is my understanding that this is possible on .NET, byt that it is not possible in WinRT (Win8, Win81) but was reintroduced in UWP. 

I build this understanding from the following:
- [MSDN: Using SO_REUSEADDR and SO_EXCLUSIVEADDRUSE](https://msdn.microsoft.com/en-us/library/ms740621%28VS.85%29.aspx?f=255&MSPPError=-2147217396)
- [MSDN Social Forum: How to reuse port for datagramsocket ?](https://social.msdn.microsoft.com/Forums/en-US/2e4cc209-843c-484e-982f-d434d663eca0/how-to-reuse-port-for-datagramsocket-?forum=winappswithhtml5)

I agree that this should be an option. 

I'll have a look at merging with the 2.0.0pre. Do you have any estimate about when this version might come out?

I've been looking quite a bit at you code over the weekend. Nice work! It's clear to see that a lot of work have been put into this library!

On another note: I was thinking maybe to do a lite fork that only targets Windows 10, Android and iOS latest editions - for simplicity. Also, I really like using Reactive Extensions so was also thinking this would replace Events (i.e. sockets accepted and messages received). Rx is not for everyone - a matter of taste and habit I think. Anyhow, how would you feel about SocketLite.PCL living side-by-side the real thing?

rdavisau commented:
Interesting links - I am surprised to find that you cannot do it on WinRT, but so be it. In that case, I'm happy to caveat that piece of functionality in the README. As an aside, I wonder if shared multicast is a way to get around the no IPC restrictions on WinRT (I assume they still exist in UWP).

Anyway - I think it is fair to remove support for legacy platforms from sockets-for-pcl. Win8 can definitely go, I'm just not sure about Win8.1. With iOS, the classic API could be dropped, Old MonoMac versions could be dropped and it would result in a simpler project as you say. Android can probably stay the same. Maybe a final 1.x release that incorporates some of the core features planned for 2.xx while keeping the full set of legacy platforms, then a 2.xx release with all the new features that only supports the newer platforms would be a good approach here?

I am a big fan of Rx (and btw if you are doing mobile and MVVM or cross platform I would recommend you check out [ReactiveUI](https://github.com/reactiveui/ReactiveUI) for an Rx inspired approach), but as an unopinionated library, sockets-for-pcl aims to ship without dependencies and without driving the user down a specific path. I do like the idea of Rx friendly sockets; would you consider creating a package that wraps the existing implementations with Rx-based alternatives? You can see for example in the sockethelpers project, aimed as a more opinionated library (note I haven't touched in a long while) that the [JsonProtocolMessenger](https://github.com/rdavisau/sockethelpers-for-pcl/blob/master/SocketHelpers/SocketHelpers/Messaging/JsonProtocolMessenger.cs) wraps a `TcpSocketClient` and provides an Rx interface (and also handles JSON serialization, deserialization , disconnection detection etc. for the user). If we drop legacy platforms, an Rx extension package this might be a more maintainable option - a fork that changes the event mechanisms might make the sharing changes back and forth more difficult in the long term. But this is just me thinking aloud so interested in your views.

1iveowl commented:
Have a look here: [SocketLite.PCL](https://github.com/1iveowl/SocketLite.PCL). It's a private library for now, but I made you a collaborator so you have access. 

I didn't create a wrap. I refactored it internally and externally for Rx. I also made some other changes. Specifically the Disconnect methods I made synchronous. I don't think they need to be async and if they do people cal always wrap them in a TaskRun themselves. I think it is cleaner that way. There was also some things regarding garbages collection that I really didn't understand, but I guess it could be related to legacy stuff?

I also removed the all comments This is probably just a matter of personal style, but a code base of this relative modest size I think is really much easier to read, understand and maintain without all the "green stuff".

My apologies if I got carried away (it was fun to do :-)), and I hope we can put it to use somehow beyond making a seperate Rx fork?
